substantial/types.py

```diff
@@ -75,23 +79,49 @@ def linear(self, retries_left: int) -> timedelta:
         )
 
 
+@dataclass
+class CompensationStrategy:
+    compensate_fn: Optional[Callable[[], Any]] = None
+    max_compensation_attemps: int = 3
+
+
+class CompensationFailed(BaseException):
+    def __init__(
+        self,
+        original_error: Exception,
+        compensation_error: Optional[Exception] = None,
+    ):
+        self.original_error = original_error
+        self.compensation_error = compensation_error
+
+
 @dataclass
 class ValueEval:
     lambda_fn: Callable[[], Any]
-    timeout: Union[int, None]
+    timeout: Union[float, None]
     retry_strategy: Union[RetryStrategy, None]
-
-    async def exec(self, ctx, save_id, counter) -> Any:
+    compensate_fn: Optional[Callable[[], Any]] = None
+    max_compensation_attempts: int = 3
+
+    async def exec(
+        self,
+        ctx,
+        save_id,
+        counter,
+        compensation_stack: List[Callable[[], Any]] = [],
+    ) -> Any:
```

Comment:
I suggest against this, the default argument in python is the same for ALL callers for mutable objects.
For example, what if there are two workflows, will they share the same compensate value?

Here I believe you can test that directly by triggering a replay on a workflow (the compensation stack will grow unexpedly).

Reply:
Can, give me what u expect…

Reply:
> The default argument in python is the same for ALL callers for mutable objects
Suppose you have two worklows, since the array is shared, they will append concurrently on the *same* thing and that's a bug.

Reply:
> The default argument in python is the same for ALL callers for mutable objects

Suppose you have two worklows, since the array is shared, they will append concurrently on the *same* thing and that's a bug.
```python
class Baz:
    def __init__(self, p=[]):  # Shared mutable default
        p.append(1)
        print(p)

class Foo:
    def bar(self):
        return Baz()

Foo().bar()
Foo().bar()
Foo().bar() 
```

Reply:
I also invite you to think of how the implementation would look like on metatype as the same thing will be done there anyways.  Also, even if the shared array were intentional (maybe after adding some memo mechanism to distinguish callers) this wouldn't have worked either in Rust or Typescript (it's a python-only "feature").